quinn/src/lib.rs

```diff
@@ -712,28 +709,22 @@ impl Connection {
     /// # Panics
     /// - If called more than once on handles to the same connection
     // FIXME: Infallible
-    pub fn close(&self, error_code: u16, reason: &[u8]) -> impl Future<Item = (), Error = ()> {
-        let (send, recv) = oneshot::channel();
-        {
-            let endpoint = &mut *self.0.endpoint.borrow_mut();
+    pub fn close(&self, error_code: u16, reason: &[u8]) {
```

Comment:
Should this consume `self`?

Reply:
It actually did in the past. Because stable futures make it so difficult to get anything done without taking ownership, `Connection` became `Clone`, with the consequence that consuming `self` doesn't accomplish much. We could get rid of that impl and encourage people to use `Rc` manually, but then closing a connection will require `Rc::try_unwrap`, which makes me worry about ergonomics.

Reply:
I would just argue that even if consuming `self` is somewhat pointless (because it's trivial and cheap to `clone()`), there might still be a benefit in terms of documenting the expected behavior and communicating how the system is supposed to work.

Reply:
(In this case, the user couldn't just call `close()` on it and continue running as normal, they'd have to have ownership of the `Endpoint`, detaching it from its owner.)